circuit_knitting/cutting/cut_finding/cco_utils.py

```diff
@@ -0,0 +1,161 @@
+    cut_types = interface.cut_type
+    qc_cut = QuantumCircuit(num_qubits)
+    for k, op in enumerate([cut_circuit for cut_circuit in cut_circuit_list]):
+        if cut_types[k] is None:  # only append gates that are not cut.
+            op_name = op.name
+            op_qubits = op.qubits
+            op_params = op.params
+            inst = Instruction(op_name, len(op_qubits), 0, op_params)
+            qc_cut.append(inst, op_qubits)
+    return qc_cut
+
+
+def select_search_engine(
+    stage_of_optimization: str,
+    optimization_settings: OptimizationSettings,
+    search_space_funcs: SearchFunctions,
+    stop_at_first_min: bool = False,
+) -> BestFirstSearch:
+    """Select the search algorithm to use.
+
+    In this release, the main search engine is always Dijkstra's
+    best first search algorithm. Note however that there is also
+    :func:``greedy_best_first_search``, which is used to warm start
+    the search algorithm. It can also provide a solution should the
+    main search engine fail to find a solution given the constraints
+    on the computation it is allowed to perform.
+    """
+    engine = optimization_settings.get_engine_selection(stage_of_optimization)
+
+    if engine == "BestFirst":
+        return BestFirstSearch(
+            optimization_settings,
+            search_space_funcs,
+            stop_at_first_min=stop_at_first_min,
+        )
+
+    else:
+        raise ValueError(f"Search engine {engine} is not supported.")
+
+
+def greedy_best_first_search(
+    state: DisjointSubcircuitsState,
+    search_space_funcs: SearchFunctions,
+    *args: CutOptimizationFuncArgs,
+) -> None | DisjointSubcircuitsState:
+    """Perform greedy best-first search using the input starting state and the input search-space functions.
+
+    The resulting goal state is returned, or None if a deadend is reached. Any additional input argumnets
+    are passed as additional arguments to the search-space functions.
+    """
+    search_space_funcs.goal_state_func = cast(
+        Callable, search_space_funcs.goal_state_func
+    )
+    search_space_funcs.cost_func = cast(Callable, search_space_funcs.cost_func)
+    search_space_funcs.next_state_func = cast(
+        Callable, search_space_funcs.next_state_func
+    )
+
+    if search_space_funcs.goal_state_func(state, *args):
+        return state
+
+    best = min(
+        [
+            (search_space_funcs.cost_func(next_state, *args), k, next_state)
+            for k, next_state in enumerate(
+                search_space_funcs.next_state_func(state, *args)
+            )
+        ],
+        default=(None, None, None),
+    )
+
+    if best[-1] is not None:
+        return greedy_best_first_search(best[-1], search_space_funcs, *args)
+
+    # The else block below covers a rare edge case
+    # which needs a clever circuit to get tested.
+    # Excluding from test coverage for now.
+    else:  # pragma: no cover
+        return None
```

Comment:
I clarified the comment and suggest using a guarded-if pattern, following e.g. https://medium.com/lemon-code/guard-clauses-3bc0cd96a2d3, to make the normal case go straight through and have the exceptional case covered by the `if` statement.

```suggestion
    if best[-1] is None:  # pragma: no cover
        # This covers a rare edge case.
        # We have so far found no circuit that triggers it.
        # Excluding from test coverage for now.
        return None

    return greedy_best_first_search(best[-1], search_space_funcs, *args)
```

Reply:
This is a good suggestion, thank you.

Reply:
'Tis done :  [f64ca8a](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/pull/471/commits/f64ca8a37aae7f645f2af31031b0b611e1921f3a)